Add configuring Bitbucket PR comments, update GitLab MR comments, additional edits and broken link fixes, add level 6 header support (#915)
returntocorp/semgrep-docs: docs/semgrep-app/notifications.md

1. In Bitbucket, select the repository where you want Semgrep to run in CI.
1. Click **Repository settings** > **Repository variables**.
1. Create a new repository variable by entering **SEMGREP_APP_TOKEN** into the **Name** field, and then paste the value of the token in the **Value** field.
1. Enable the **Secured** option, and then click **Add**.

#### Adding Semgrep to your Bitbucket CI pipeline for PR comments

To receive comments on PRs, add the following code to your `bitbucket-pipelines.yml` file:

```yaml
image: atlassian/default-image:latest

pipelines:
  pull-requests:
    '**':
      - step:
          name: 'Run Semgrep scan in PR branch'
          image: returntocorp/semgrep
          script:
            - export SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN
            - export SEMGREP_APP_URL="https://semgrep.dev"
            - export SEMGREP_BASELINE_REF="origin/main"
            - git fetch origin "+refs/heads/*:refs/remotes/origin/*"
            - export BITBUCKET_TOKEN=$PAT
            - semgrep ci

> ```

Author: @pangngan let me know what we need to edit here to make this scan the main branches also. :) I was looking here - https://semgrep.dev/docs/semgrep-ci/sample-ci-configs/#sample-bitbucket-pipelines-configuration-snippet should we update the code snippet too?

Reviewer: The reason why the pipeline event trigger is `pull-requests` (line 393) is because you can't generate a PR comment on a MAIN (or default) branch. It's in the name of the comment: PR comment.

To enable GitLab merge request comments, follow these steps: 

1. Create an API token in GitLab by going to [Profile > Access Tokens](https://gitlab.com/-/profile/personal_access_tokens), and then add a token with `api` scope.
1. Copy the token created in the previous step.
1. Navigate to **Your repository** >  **Settings** > **CI/CD**. The URL of the page where you are ends with: `/username/project/-/settings/ci_cd`.
1. Under **Variables** click **Expand**, and then click **Add variable**.
1. Enter `PAT` in the **Key** field and paste the token value from the step two to the **Value** field.
1. Select **Mask variable** checkbox option, and then clear the **Protect variable** checkbox option.
1. Update your `.gitlab-ci.yml` file with variable `GITLAB_TOKEN` and value `$PAT`. Refer to the following example:
```yaml
semgrep:
  # A Docker image with Semgrep installed.
  image: returntocorp/semgrep
  # Run the "semgrep ci" command on the command line of the docker image.
  script: semgrep ci

  rules:
  # Scan changed files in MRs, (diff-aware scanning):
  - if: $CI_MERGE_REQUEST_IID

>   # Scan mainline (default) branches and report all findings.
>   - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH

Author: ```suggestion
```

1. Select **Scan new project**, and then **Run Scan in CI**.
1. Click **Create new API token**, and then copy the value of the new API token.
1. In GitLab, add the token to your project by going to **Settings** > **CI/CD**.
1. Under **Variables** click **Expand**, and then click **Add variable**.
1. Enter `SEMGREP_APP_TOKEN` in the **Key** field and paste the token value in the **Value** field.
1. Select **Mask variable** checkbox option, and then clear the **Protect variable** checkbox option.
1. Click **Add variable**.

For more configuration options, see [GitLab CI Sample](/semgrep-ci/sample-ci-configs/#gitlab-ci).

:::note
GitLab MR comments are only available to logged-in Semgrep Cloud Platform users.
:::

### Bitbucket pull request comments

To configure Semgrep PR comments in your Bitbucket PRs, follow the steps described in the subsections below.

#### Creating a personal access token

Create a personal access token to authenticate to the Bitbucket API. There are two ways in which you can create personal access tokens depending on the Bitbucket plan you use:

> - **Workspace access token**: If you use the Bitbucket Cloud Premium plan, you can create a workspace access token. This option saves time because you can create one access token for all repositories in the workspace. See [Creating a workspace access token](/semgrep-app/notifications/#creating-and-adding-a-workspace-access-token).

Author: ```suggestion
- **Workspace access token**: If you use the Bitbucket Cloud Premium plan, you can create a workspace access token. This option saves time because you can create one access token for all repositories in the workspace. With one access token for all repositories in the workspace, you can bulk-onboard more repositories at once from a whole workspace. See [Creating a workspace access token](/semgrep-app/notifications/#creating-and-adding-a-workspace-access-token). However, you can also use the option of a repository access token, see below for more information to onboard repositories one by one.
```

> - **Workspace access token**: If you use the Bitbucket Cloud Premium plan, you can create a workspace access token. This option saves time because you can create one access token for all repositories in the workspace. See [Creating a workspace access token](/semgrep-app/notifications/#creating-and-adding-a-workspace-access-token).

Author: @pangngan great point to mention!

> - **Repository access token**: If you are **not** using the Bitbucket Cloud Premium plan, you have to create a separate repository access token for each repository where you want to use Semgrep. See [Creating a repository access token](/semgrep-app/notifications/#creating-and-adding-a-repository-access-token).

Author: ```suggestion
- **Repository access token**: If you do **not** have the Bitbucket Cloud Premium plan, create a separate repository access token for each repository where you want to use Semgrep. This configuration option is also useful if you have the Bitbucket Cloud Premium plan, but prefer to onboard repositories one by one instead of bulk onboarding. See [Creating a repository access token](/semgrep-app/notifications/#creating-and-adding-a-repository-access-token).
```

:::

Create a workspace access token in Bitbucket (only available if you have a Bitbucket Cloud Premium plan). Follow the instructions in [Create a Workspace Access Token](https://support.atlassian.com/bitbucket-cloud/docs/create-a-workspace-access-token/). Alternatively, follow these steps to create a workspace access token:

1. In the Bitbucket workspace, click the cogwheel icon next to your profile icon in the top right.
1. Click **Workspace settings** > **Access tokens**.
1. Click **Create Workspace Access Token**.
1. Create a token name. Note: As Semgrep comments to your PRs, use **Semgrep** as the name for this token.
1. Under **Scopes**, select the **Read**, and **Write** permissions for pull requests.
1. Click **Create**.
1. Copy the value of the workspace access token, and then click **Close**.
1. Click **Workspace variables**.
1. Create a new workspace variable by entering **PAT** into the **Name** field, and then paste the workspace access token in the **Value** field.
1. Enable the **Secured** option, and then click **Add**.

> Create and add a `SEMGREP_APP_TOKEN` to establish the communication between your workspace and the Semgrep Cloud Platform by following these steps:

Reviewer: Suggest to remove lines 357 to 363 and to do the same for the subsequent section.

Author: @s-santillan do you mean to delete Creating and adding a repository access token? But users **without** Bitbucket Cloud Premium cannot do otherwise. The only way they can get our PR comments is by adding a repository access token (contrary to Bitbucket Cloud Premium users who can add a workspace access token).

Why do want to delete this section? My main concern is - the goal should be that users successfully get Semgrep messages to their PRs. Do you have an alternative here?

Author: Or do you mean the part about the `SEMGREP_APP_TOKEN` in the subsequent section?

Each part of the docs should be as independent as possible. In some sections, you can imagine that our users went just to this page to see how to enable the comments. Their sole goal is to have PR or MR comments, nothing more. Maybe we can make the `SEMGREP_APP_TOKEN` part optional? Also, how to add the token is different in every CMS... so documenting this seems to make sense.

Reviewer: By the time the users already get to adding comments, they will have already set up the SEMGREP_APP_TOKEN. 
Flow is like this:
1. Users add repo through Semgrep Cloud PlatformP -> THIS step adds the SEMGREP_APP_TOKEN. There is no way for a user to scan a repo with Semgrep Cloud Platform without it. **This is true for all plans, in all SCMs**, whether it is GitHub's free or paid tier or Bitbucket's free or paid tier. It establishes the connection between the repo and SCP. 
2. Users enable various notifications (PR/MR comments, Slack, etc.)

I think the reason why SEMGREP_APP_TOKEN was not "clear" is because onboarding for a BitBucket repo hasn't been added in the docs. You can however see the SEMGREP_APP_TOKEN being added in various other guides.

Reviewer: Here's the guide for setting up a connection in CI: https://semgrep.dev/docs/semgrep-ci/running-semgrep-ci-with-semgrep-app/#ci-providers-listed-within-semgrep-cloud-platform-such-as-github-actions-gitlab-cicd-jenkins

Reviewer: I'll answer this first:

> Note: I am also not sure that a header **CI providers listed within Semgrep Cloud Platform (such as GitHub Actions, GitLab CI/CD, Jenkins)** is easily understood as the place to search for: Creating SEMGREP_APP_TOKEN.

The parent section is [Setting up the CI job and Semgrep Cloud Platform Connection](https://semgrep.dev/docs/semgrep-ci/running-semgrep-ci-with-semgrep-app/#setting-up-the-ci-job-and-semgrep-cloud-platform-connection). It subsections are "CI providers listed within Semgrep Cloud Platform" and "Other CI providers" because the steps differ between those two categories.

Does that answer your question?

Author: The admonition is great, putting thumbs up on the last two comments. My main question is: How to actually copy the token from the Cloud Platform so users do not need to create a new one every time? @s-santillan

Author: Mentioned above: @s-santillan ![Screenshot 2023-04-03 at 16 41 24](https://user-images.githubusercontent.com/26669307/229543689-feb10a7c-6ecd-4e94-9edb-1db94a1145b1.png)

Reviewer: By design, you (admin, member, etc.) can't copy a token after it's been created. Just make a new token if you need to. "If you already have a token configured" refers to the act of copying the token into the CI provider as a secret, which configures the connection.

1. Enable the **Secured** option, and then click **Add**.

To complete the configuration, follow the [Adding Semgrep to your Bitbucket CI pipeline for PR comments](#adding-semgrep-to-your-bitbucket-ci-pipeline-for-pr-comments) section.

#### Adding Semgrep to your Bitbucket CI pipeline for PR comments

To receive comments on PRs, commit the following code to your `bitbucket-pipelines.yml` file:

```yaml
image: atlassian/default-image:latest

pipelines:
  branches:
    main:
        - step:
            name: 'Run Semgrep full scan with main branch'
            image: returntocorp/semgrep
            script:
              - export SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN
              - semgrep ci
  pull-requests:
    '**':
      - step:
          name: 'Run Semgrep diff scan with PR branch'
          image: returntocorp/semgrep
          script:
            # Set SEMGREP Variables
            - export SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN
            - export SEMGREP_BASELINE_REF="origin/main"
            - git fetch origin "+refs/heads/*:refs/remotes/origin/*"
            - export BITBUCKET_TOKEN=$PAT
            - semgrep ci

> ```

Author: @s-santillan just tagging you here, I was not sure if you want to add this to `/src/components/code_snippets/_bitbucket-semgrep-app-sast.mdx` and add more comments. :)
